botocore/configprovider.py

```diff
@@ -851,3 +858,145 @@ def provide(self):
+                self._client_name,
+                location,
+            )
+
+            endpoint_url = getattr(self, f"_get_endpoint_url_{location}")()
+
+            if endpoint_url:
+                logger.info(
+                    "Found endpoint for %s via: %s.",
+                    self._client_name,
+                    location,
+                )
+                return endpoint_url
+
+        logger.debug("No configured endpoint found.")
+        return None
+
+    def _get_endpoint_url_environment_service(self):
+        return EnvironmentProvider(
+            name=self._get_service_env_var_name(), env=self._environ
+        ).provide()
+
+    def _get_endpoint_url_environment_global(self):
+        return EnvironmentProvider(
+            name="AWS_ENDPOINT_URL", env=self._environ
+        ).provide()
+
+    def unhyphenize_service_name(self, service_name):
+        return service_name.replace('-', ' ')
+
+    def _get_service_id(self, client_name):
+        client_name = utils.handle_service_name_alias(client_name)
+
+        client_name = utils.CLIENT_NAME_TO_HYPHENIZED_SERVICE_ID_OVERRIDES.get(
+            client_name, client_name
+        )
+
+        return self.unhyphenize_service_name(client_name)
+
+    def _get_endpoint_url_config_service(self):
+        snakecase_service_id = self._snakecase_service_id(
+            self._get_service_id(self._client_name)
+        ).lower()
+        return (
+            self._get_services_config()
+            .get(snakecase_service_id, {})
+            .get('endpoint_url')
+        )
+
+    def _get_endpoint_url_config_global(self):
+        return self._scoped_config.get("endpoint_url")
+
+    def _snakecase_service_id(self, service_id):
+        return service_id.replace(" ", "_")
+
+    def _get_service_env_var_name(self):
+        transformed_service_id_env = self._snakecase_service_id(
+            self._get_service_id(self._client_name)
+        ).upper()
+        return f"AWS_ENDPOINT_URL_{transformed_service_id_env}"
+
+    def _get_services_config(self):
+        if "services" not in self._scoped_config:
+            return {}
+
+        section_name = self._scoped_config["services"]
+        services_section = self._full_config.get("services", {}).get(
+            section_name
+        )
+
+        if not services_section:
+            error_msg = (
+                f'The profile is configured to use the services '
+                f'section but the "{section_name}" services '
+                f"configuration does not exist."
+            )
+            raise InvalidConfigError(error_msg=error_msg)
+
+        return services_section
+
+    def __deepcopy__(self, memo):
+        return ConfiguredEndpointProvider(
+            full_config=copy.deepcopy(self._full_config, memo),
+            scoped_config=copy.deepcopy(self._scoped_config, memo),
+            client_name=copy.deepcopy(self._client_name),
+            environ=copy.deepcopy(self._environ),
+        )
```

Comment:
A couple of things here. Do we need a `deepcopy` implementation? I'm pretty sure all of these values are native Python data types so the custom logic isn't necessary. If we do though, I have a couple of suggestions.

1. `client_name` is a string and therefore immutable. Copying it is pointless and unnecessary.
2. We may want to account for recursive calls here by doing some work with `memo` instead of just handing it off to `copy.deepcopy`. Something like the following could be added:

```python
if memo is None:
    memo = {}

if id(self) in memo:
    return memo[id(self)]

new_instance = ConfiguredEndpointProvider(...)
memo[id(new_instance)] = new_instance
return new_instance

Reply:
An additional nit: constructing the copy with `self.__class__` is probably a bit more maintainable.

Reply:
Existing config providers that have constant `str` values (and other native Python datatypes) also explicitly deep copy them. Example:

https://github.com/boto/botocore/blob/a447b7a9e849e216e31ca4aa616fba6e70b2b988/botocore/configprovider.py#L831

Reply:
That doesn't mean they're right 😃 I would still remove the copy from the string since it isn't doing anything.

Reply:
OK! I tested out that the class is deep-copyable without this implementation, so I removed it and added a unit test to prevent regression.